apisix/kms.lua

```diff
@@ -0,0 +1,182 @@
+--
+-- Licensed to the Apache Software Foundation (ASF) under one or more
+-- contributor license agreements.  See the NOTICE file distributed with
+-- this work for additional information regarding copyright ownership.
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+
+local require   = require
+local core      = require("apisix.core")
+local string    = require("apisix.core.string")
+
+local find      = string.find
+local sub       = string.sub
+local upper     = string.upper
+local type      = type
+local pcall     = pcall
+local ipairs    = ipairs
+local error     = error
+
+local _M = {
+    version = 0.1,
+}
+
+
+local KMS_PREFIX = "$KMS://"
+local kmss
+
+local function check_kms(conf)
+    local idx = find(conf.id or "", "/")
+    if not idx then
+        return false, "no kms id"
+    end
+    local service = sub(conf.id, 1, idx - 1)
+
+    local ok = pcall(require, "apisix.kms." .. service)
+    if not ok then
+        return false, "kms service not exits, service: " .. service
+    end
+
+    return core.schema.check(core.schema["kms_" .. service], conf)
```

Comment:
It would be better to let each `"apisix.kms." .. service` module define its schema inside. Just like a plugin or a service discovery.